sdk/storage/Azure.Storage.Internal.Avro/src/AvroParser.cs

```diff
@@ -0,0 +1,388 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+#pragma warning disable SA1402 // File may only contain a single type
+
+namespace Azure.Storage.Internal.Avro
+{
+    internal static class AvroParser
+    {
+        public static async Task<byte[]> ReadFixedBytesAsync(
+            Stream stream,
+            int length,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] data = new byte[length];
+            int start = 0;
+            while (length > 0)
+            {
+                int n = async ?
+                    await stream.ReadAsync(data, start, length, cancellationToken).ConfigureAwait(false) :
+                    stream.Read(data, start, length);
+                start += n;
+                length -= n;
+
+                // We hit the end of the stream
+                if (n <= 0)
+                    return data;
+            }
+            return data;
+        }
+
+        private static async Task<byte> ReadByteAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] bytes = await ReadFixedBytesAsync(stream, 1, async, cancellationToken).ConfigureAwait(false);
+            return bytes[0];
+        }
+
+        // Stolen because the linked references in the Avro spec were subpar...
```

Comment:
Might want to revisit that comment. End of day it's legit under their license.

Reply:
Removed the comment.